server/app/middlewares/websocket_backend.rb

```diff
@@ -236,19 +267,30 @@ def our_version
 
   # @param [Faye::WebSocket] ws
   # @param [HostNode] node
-  def handle_invalid_agent_version(ws, node)
-    node.set(connected: false, last_seen_at: Time.now.utc)
-    EventMachine::Timer.new(1) do
-      ws.close(4010) if ws
-    end
+  def handle_invalid_agent_version(ws, node, version)
+    send_master_info(ws)
+    ws.close(4010, "agent version #{version} is not compatible with server version #{Server::VERSION}")
+  end
+
+  # Send master_info RPC notification directly, without looping through the normal RPC mechanisms
+  #
+  # @param [Faye::Websocket] ws
+  def send_master_info(ws)
+    # symbols in RPC parameters are implicitly converted into strings by MongoPubsub
+    send_rpc_notify(ws, '/agent/master_info', {'version' => Server::VERSION})
   end
 
   # @param [Faye::Websocket] ws
+  def send_rpc_notify(ws, method, *params)
+    send_message(ws, [2, method, params])
+  end
+
+  # Must be called in EM thread.
+  #
+  # @param [Faye::Websocket] ws
   # @param [Array] message
   def send_message(ws, message)
-    EM.next_tick {
-      ws.send(MessagePack.dump(message).bytes)
-    }
+    ws.send(MessagePack.dump(message).bytes)
```

Comment:
Tested the new `handle_invalid_agent_version` logic to make sure it  still works:

```
core@core-02 ~/share $ cat VERSION 
1.2.2.dev
$ cat kontena.env 
KONTENA_VERSION=1.2.2.dev
$ docker-compose run agent
...
I, [2017-05-26T12:22:04.557960 #1]  INFO -- Kontena::WebsocketClient: connecting to master at ws://192.168.66.1:9292
I, [2017-05-26T12:22:04.680311 #1]  INFO -- Kontena::WebsocketClient: connection established
E, [2017-05-26T12:22:04.685189 #1] ERROR -- Kontena::WebsocketClient: master does not accept our version (1.2.2.dev): agent version 1.2.2.dev is not compatible with server version 1.3.0.pre1
D, [2017-05-26T12:22:04.685908 #1] DEBUG -- Kontena::RpcServer: rpc notification: Kontena::Rpc::AgentApi#master_info [{"version"=>"1.3.0.pre1"}]
D, [2017-05-26T12:22:04.687231 #1] DEBUG -- Kontena::Workers::NodeInfoWorker: publishing node information
I, [2017-05-26T12:22:04.686754 #1]  INFO -- Kontena::Workers::LogWorker: stopped log streaming
I, [2017-05-26T12:22:04.696142 #1]  INFO -- Kontena::Workers::LogWorker: started log streaming
Shutting down ...
I, [2017-05-26T12:22:04.854881 #1]  INFO -- Kontena::Workers::ContainerHealthCheckWorker: stopped to check status from /ingress-lb.lb-2
$ cat kontena.env 
KONTENA_VERSION=1.3.0.pre1
```

It seems to work now, without the magic 1s timer for `ws.close` that was there before.